src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -137,17 +135,21 @@ protected override Task InvokeResultAsync(IActionResult result)
 
         private object CreateInstance()
         {
-            if (HasPageModel)
+            if (_actionDescriptor.ModelTypeInfo != _actionDescriptor.PageTypeInfo)
```

Comment:
This has become way less self-documenting. This needs more description about what kinds of cases are possible and why they are important.